### PR TITLE
Update MerchWeek and deprecate methods from MerchCalendar

### PR DESCRIPTION
**Pull request by chylauSF**
## Problem
MerchWeek class needs to be updated to return Fiscal Year dates, months, quarters, and etc, based on what current date it is.

## Solution
MerchWeek needs to know what calendar it needs to base its computation on, either using the RetailCalendar or the StitchFixFiscalCalendar.  Depending on what date the MerchWeek class consumes, it can return the appropriate information based on what calendar it receives.  If the user does not specify the calendar, it defaults to the RetailCalendar.

We also saw that in the util.rb there were some methods that we needed to use for the StitchFixFiscalCalendar to use, but since the util class was built based on the RetailCalendar, we had to copy some of the methods and add it to the StitchFixFiscalCalendar.  MerchCalendar was originally built for calculating dates in terms of Retail Calendar, but now we want to break that assumption and use for any calendar, fiscal, or whatever we need to add to this gem. 

Things I need help on:
- Should I make the RetailCalendar completely independent of using the util class? 
- Any sort of re-arranging of tests?  Any more tests to add?  Refactoring tests?
-  In MerchWeek.rb:13, I am not sure if I should expose the Calendar attribute
-  Let me know if I document the code correctly?
- What other things I should write in the ReadMe?

### Pre-deployment
- [x] Update the ReadMe to better reflect the current purpose of the gem
- [ ] Adding more coverage based on what coveralls says
- [x] Awaiting Reviews

### Post deployment
- [ ] Bump gem version to 0.3.0 and release to rubygems.

## Reviewers:
@jonworek  @davidamcclain @sleepn247

### CODE REVIEW

**Comment by sleepn247** [comment]
sorry just realized I have to press end review for the comments to show up, that was from this morning before your latest commit

File: lib/merch_calendar/merch_week.rb [hunk 1]
```diff
@@ -9,57 +9,68 @@ class MerchWeek
...
       #   @return [Array<MerchWeek>]
       # @overload find(year, julian_month, week_number)
       #   @param year [Fixnum] the merch year to locate
       #   @param julian_month [Fixnum] the month to find merch months for
       #   @param week_number [Fixnum] the specific week number.
+      #   @param options [Hash] options to set your calendar, if none it will default to RetailCalendar
       #   @return [MerchWeek] the specific merch week
-      def find(year, julian_month, week_number=nil)
+      def find(year, julian_month, week_number=nil, options={})
+        calendar = options.fetch(:calendar, RetailCalendar.new)
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/merch_week.rb]: I am not sure what you're asking here?  Calendar is refers to the different calendars that we need to use for the rest of the methods.  Maybe I should've put something in the documentation above to make it clearer?

File: lib/merch_calendar/merch_week.rb [hunk 2]
```diff
@@ -9,57 +9,68 @@ class MerchWeek
...
-
-
     # Pass in a date, make sure it is a valid date object
     # @private
     def initialize(date, options = {})
       @date = date
-
+      
+      #defaults to Retail Calendar if no other calendar is defined
+      @calendar = options.fetch(:calendar, RetailCalendar.new)
```
Comment (sleepn247) [inline lib/merch_calendar/merch_week.rb]: here's the initialized calendar here, referring to prev comment

File: lib/merch_calendar/retail_calendar.rb [hunk 3]
```diff
@@ -93,12 +125,97 @@ def end_of_quarter(year, quarter)
...
+      end
+    end
+    ### IMPORTANT COVER TESTS PLEASE!!!
+    
+    # Converts a merch month to the correct julian month
+    #
+    # @param  [Fixnum] the merch month to convert
+    # @return [Fixnum] the julian month
+    def merch_to_julian(merch_month)
+      if merch_month > 12 || merch_month <= 0
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/retail_calendar.rb]: What are some good places to put this check in?

File: lib/merch_calendar/retail_calendar.rb [hunk 4]
```diff
@@ -112,5 +229,50 @@ def merch_months_in(start_date, end_date)
...
+      (1..weeks).map do |week_num|
+        week_start = start_date + ((week_num - 1) * 7)
+        week_end = week_start + 6
+
+        MerchWeek.new(week_start, { 
+          start_of_week: week_start, 
+          end_of_week: week_end, 
+          week: week_num, 
+          calendar: RetailCalendar.new
+        })
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/retail_calendar.rb]: Yep it is!

File: lib/merch_calendar/retail_calendar.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (sleepn247) [reply lib/merch_calendar/retail_calendar.rb]: I'm kind of two minds about this, the "nice way" is to make sure that we have this kind of error checking through all of our public interfaces. The "wild west" way is to say, if you give me a bad month, screw you. I think that when it's unevenly applied the danger is that we expect them to perform in a "nice way."

File: lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb [hunk 5]
```diff
@@ -87,33 +135,132 @@ def end_of_month(year, merch_month)
...
+    # @return [Fixnum] the number of weeks within the fiscal year
     def weeks_in_year(year)
       ((start_of_year(year + 1) - start_of_year(year)) / 7).to_i
     end
+
+    # Given any julian date it will return what Fiscal Year it belongs to
+    #
+    # @param [Date] the julian date to convert to its Fiscal Year
+    # @return [Fixnum] the fiscal year that the julian date falls into
+    def merch_year_from_date(date)
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb]: The purpose of this method is that it receives a date object and it returns a merch year that the date object lands in.  

So for a Retail Calendar if I gave it the Date: `1/29/17` the merch_year should be `2017`.  

If I gave it a Date object of ` 1/1/18` the merch_year should be `2017` 

Let me know if you have a better name for it!

File: lib/merch_calendar/merch_week.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (sleepn247) [reply lib/merch_calendar/merch_week.rb]: oops! I read this wrong

File: lib/merch_calendar/merch_week.rb (hunk 2 above)
Comment (chylauSF) [reply lib/merch_calendar/merch_week.rb]: Yes it's for backwards compatibility reasons.  Currently we have some parts of FT that uses the Retail Calendar methods, and if they didn't have a Calendar class identified then we use the Retail Calendar class as a default.  

So people who are already using the MerchWeek class and didn't identified the Calendar, its okay because their default is the retail calendar and nothing breaks.

File: lib/merch_calendar/merch_week.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (sleepn247) [reply lib/merch_calendar/merch_week.rb]: the find operates on the class method level, so we need the duplication. Sorry about that.

File: spec/merch_calendar/merch_week_spec.rb [hunk 6]
```diff
@@ -40,110 +62,242 @@
...
+          expect(merch_week.year_week).to eq date_check.year_week
+        end
 
-      OpenStruct.new(date: "2014-02-01", year: 2013, month: 1,  week: 4, quarter: 2, year_week: 52, start_date: "2013-02-03"),
+        it "#month" do
+          expect(merch_week.month).to eq date_check.month
+        end
 
-      OpenStruct.new(date: "2015-02-01", year: 2015, month: 2,  week: 1, quarter: 3, year_week: 1, start_date: "2015-02-01"),
+        it "#quarter" do
```
Comment (jonworek) [inline spec/merch_calendar/merch_week_spec.rb]: `it "#quarter" do` and all the similarly named blocks don't follow rspec conventions.

`describe "#quarter" do` is more correct.  But I think you'll have to rework this whole section, starting at the parent `context` on line 215 (which should probably be a `describe` instead)

File: spec/merch_calendar/merch_week_spec.rb [hunk 7]
```diff
@@ -40,110 +62,242 @@
...
+        end
+
+        it "#year" do
+          expect(merch_week.year).to eq date_check.year
+        end
+      end
+    end
+  end
+
+  context "logic for Stitch Fix Fiscal Calendar" do
```
Comment (jonworek) [inline spec/merch_calendar/merch_week_spec.rb]: This should also be a `describe`.

File: spec/merch_calendar/retail_calendar_spec.rb [hunk 8]
```diff
@@ -57,6 +60,35 @@
       expect(subject.end_of_quarter(2019, 4)).to eq Date.new(2020, 2, 1)
     end
   end
+  
+  describe "#quarter" do
+    it "returns the correct date" do
```
Comment (jonworek) [inline spec/merch_calendar/retail_calendar_spec.rb]: This isn't correct.  `it "returns the correct quarter number"` would make more sense.

File: spec/merch_calendar/retail_calendar_spec.rb [hunk 9]
```diff
@@ -57,6 +60,35 @@
...
+      expect(subject.quarter(5)).to eq 2
+      expect(subject.quarter(7)).to eq 3
+      expect(subject.quarter(2)).to eq 1
+      expect(subject.quarter(11)).to eq 4
+    end
+  end
+  
+  describe "#season" do
+    context "returns Spring/Summer from its merch_month" do
```
Comment (jonworek) [inline spec/merch_calendar/retail_calendar_spec.rb]: `context "for merch months in the Spring/Summer season"` more accurately describes the context.

Saying "something returns something" is only valid with an `it "returns..."`.  

For all of these rspec things I'm pointing out, a good guideline is to read the test out loud to yourself.  It should make sense in plain English.

File: spec/merch_calendar/retail_calendar_spec.rb [hunk 10]
```diff
@@ -121,5 +152,84 @@
...
+      expect(subject.merch_to_julian(4)).to eq 5
+      expect(subject.merch_to_julian(5)).to eq 6
+      expect(subject.merch_to_julian(6)).to eq 7
+      expect(subject.merch_to_julian(7)).to eq 8
+      expect(subject.merch_to_julian(8)).to eq 9
+      expect(subject.merch_to_julian(9)).to eq 10
+      expect(subject.merch_to_julian(10)).to eq 11
+      expect(subject.merch_to_julian(11)).to eq 12
+      expect(subject.merch_to_julian(12)).to eq 1
+      expect { subject.merch_to_julian(13) }.to raise_error ArgumentError
```
Comment (jonworek) [inline spec/merch_calendar/retail_calendar_spec.rb]: I'd move these to a separate, `it` block, personally... `it "raises an error for invalid merch months"`

File: spec/merch_calendar/retail_calendar_spec.rb [hunk 11]
```diff
@@ -121,5 +152,84 @@
...
+    end
+  end
+  
+  describe "#weeks_for_month" do
+    context "correct number of weeks given julian month and fiscal year" do
+      it "returns 4 weeks for a 4-week month Fiscal Year 2019 for Aug" do
+        weeks = subject.weeks_for_month(2018, 2)
+        expect(weeks.size).to eq 4
+      end
+      it "returns 5 weeks for a 5-week month Fiscal Year 2019 for Sept" do
```
Comment (jonworek) [inline spec/merch_calendar/retail_calendar_spec.rb]: Please add newlines between blocks.

File: spec/merch_calendar/retail_calendar_spec.rb [hunk 12]
```diff
@@ -121,5 +152,84 @@
...
+      expect(subject.julian_to_merch(11)).to eq 10
+      expect(subject.julian_to_merch(12)).to eq 11
+      expect(subject.julian_to_merch(1)).to eq 12
+      expect { subject.julian_to_merch(13) }.to raise_error ArgumentError
+      expect { subject.julian_to_merch(0) }.to raise_error ArgumentError
+    end
+  end
+  
+  describe "#weeks_for_month" do
+    context "correct number of weeks given julian month and fiscal year" do
```
Comment (jonworek) [inline spec/merch_calendar/retail_calendar_spec.rb]: I don't think this qualifies as a `context`.   Contexts are used to group related specs together, usually indicating a precondition of all specs that fall under it.

ex...

```
describe "a login page" do
  context "user is already logged in" do
    it "tells the user they're already logging in" do
...
    it "takes the user to the 'logged in user' page" do
...
```

File: lib/merch_calendar/merch_week.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/merch_week.rb]: Ahhh okay!  I can re write it!

File: lib/merch_calendar/merch_week.rb [hunk 13]
```diff
@@ -9,57 +9,68 @@ class MerchWeek
     #
     # @!attribute [r] date
     #   @return [Date] the date for this merch week
-    attr_reader :date
+
+    # The Merch Calendar that is being represented, either Fiscal or Retail
+    #
+    # @!attribute [r] calendar
+    #   @return [Class] the calendar that determines the merch week
+    attr_reader :date, :calendar
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/merch_week.rb]: 👍

File: lib/merch_calendar/merch_week.rb [hunk 14]
```diff
@@ -9,57 +9,68 @@ class MerchWeek
...
 
     class << self
 
       # Locates the +MerchWeek+ for a given Julian date.
       #
       # @overload from_date(String)
       #   @param [String] julian_date a julian date in the format of +YYYY-MM-DD+
+      #   @param options [Hash] options to set your calendar, if none it will default to RetailCalendar
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/merch_week.rb]: 👍

File: lib/merch_calendar/merch_week.rb [hunk 15]
```diff
@@ -9,57 +9,68 @@ class MerchWeek
...
+          calendar.weeks_for_month(year, julian_month)[week_number-1]
         end
       end
 
       # Returns the +MerchWeek+ for today's date
       #
       # @return [MerchWeek]
-      def today
-        MerchWeek.from_date Date.today
+      def today(options={})
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/merch_week.rb]: 👍

File: lib/merch_calendar/merch_week.rb [hunk 16]
```diff
@@ -81,108 +92,97 @@ def year_week
     end
 
     # This returns the "merch month" number for a date
-    # Merch months are shifted by one. Month 1 is Feb
+    # Merch months are shifted by one.
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/merch_week.rb]: 👍

File: lib/merch_calendar/merch_week.rb [hunk 17]
```diff
@@ -81,108 +92,97 @@ def year_week
...
     # The julian month that this merch week falls in
     #
     # @return [Fixnum]
     def month
-      @month ||= MerchCalendar.merch_to_julian(merch_month)
+      @month ||= calendar.merch_to_julian(merch_month)
     end
 
-    # The specific quarter this week falls in
+    # The quarter that this week falls in
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/merch_week.rb]: 👍

File: lib/merch_calendar/retail_calendar.rb [hunk 18]
```diff
@@ -2,35 +2,44 @@
...
     FOUR_WEEK_MONTHS = [2, 5, 8, 11]
     FIVE_WEEK_MONTHS = [3, 6, 9, 12]
-
+    
+    # The the first date of the retail year
+    #
+    # @param [Fixnum] year - the retail year
+    # @return [Date] the first date of the retail year
     def end_of_year(year)
-      year_end = Date.new((year + 1), 1, -1) # Jan 31st
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/retail_calendar.rb]: 👍

File: lib/merch_calendar/retail_calendar.rb [hunk 19]
```diff
@@ -48,6 +57,11 @@ def start_of_month(year, merch_month)
       start
     end
 
+    # The ending date of the given merch month
+    #
+    # @param [Fixnum] year - the retail year
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/retail_calendar.rb]: Wait... does this mean I have to change every method to:
@param _name_ [Type] _description_

File: lib/merch_calendar/retail_calendar.rb [hunk 20]
```diff
@@ -93,12 +125,95 @@ def end_of_quarter(year, quarter)
...
+
+    
+    # Converts a merch month to the correct julian month
+    #
+    # @param  [Fixnum] the merch month to convert
+    # @return [Fixnum] the julian month
+    def merch_to_julian(merch_month)
+      if merch_month > 12 || merch_month <= 0
+        raise ArgumentError
+      end
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/retail_calendar.rb]: 👍

File: lib/merch_calendar/retail_calendar.rb [hunk 21]
```diff
@@ -93,12 +125,95 @@ def end_of_quarter(year, quarter)
...
+    end
+    
+    # Converts a julian month to a merch month
+    #
+    # @param  [Fixnum] the julian month to convert
+    # @return [Fixnum] the merch month
+    def julian_to_merch(julian_month)
+      if julian_month > 12 || julian_month <= 0
+        raise ArgumentError
+      end
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/retail_calendar.rb]: 👍

File: lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb [hunk 22]
```diff
@@ -87,33 +135,132 @@ def end_of_month(year, merch_month)
...
+    end
+
+    # Converts a merch month to the correct julian month
+    #
+    # @param merch_month [Fixnum] the merch month to convert
+    # @return [Fixnum] the julian month
+    def merch_to_julian(merch_month)
+      if merch_month > 12 || merch_month <= 0
+        raise ArgumentError
+      end
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb]: 👍

File: lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb [hunk 23]
```diff
@@ -87,33 +135,132 @@ def end_of_month(year, merch_month)
...
+    end
+
+    # Converts a julian month to a fiscal month
+    #
+    # @param julian_month [Fixnum] the julian month to convert
+    # @return [Fixnum] the merch month
+    def julian_to_merch(julian_month)
+      if julian_month > 12 || julian_month <= 0
+        raise ArgumentError
+      end
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb]: 👍

File: lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb [hunk 24]
```diff
@@ -87,33 +135,132 @@ def end_of_month(year, merch_month)
...
+          end_of_week: week_end, 
+          week: week_num, 
+          calendar: StitchFixFiscalYearCalendar.new 
+        })
+      end
+    end
+
+    private
+
+    def merch_month_combo_from_dates(start_date, end_date)
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb]: I am not sure how to rename it, but I'll link you to change.  I also added in the comments.

File: spec/merch_calendar/merch_week_spec.rb [hunk 25]
```diff
@@ -40,110 +62,242 @@
         it { expect{described_class.from_date("2015")}.to raise_error(ArgumentError) }
         it { expect{described_class.from_date("2015-04")}.to raise_error(ArgumentError) }
       end
+      
+      context "wants to know a date in a Fiscal Calendar" do
+        it "allows calendar to be passed and translate date to what it looks like in a FY year" do
+          mw = described_class.from_date( "2019-08-04", fiscal_calendar_options )
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply spec/merch_calendar/merch_week_spec.rb]: 👍

File: spec/merch_calendar/merch_week_spec.rb [hunk 26]
```diff
@@ -40,110 +62,242 @@
...
+    let(:fiscal_week) { described_class.from_date("2019-08-01", fiscal_calendar_options) }
 
     it ":short / default format" do
       expect(merch_week.to_s(:short)).to eq "Dec W5"
       expect(merch_week.to_s).to eq "Dec W5"
       expect("#{merch_week}").to eq "Dec W5"
+      
+      expect(fiscal_week.to_s(:short)).to eq "Jul W5"
+      expect(fiscal_week.to_s).to eq "Jul W5"
+      expect("#{fiscal_week}").to eq "Jul W5"
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply spec/merch_calendar/merch_week_spec.rb]: 👍

File: spec/merch_calendar/retail_calendar_spec.rb (hunk 8 above)
Comment (chylauSF) [reply spec/merch_calendar/retail_calendar_spec.rb]: 👍

File: lib/merch_calendar/retail_calendar.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/retail_calendar.rb]: 👍

File: lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb [hunk 27]
```diff
@@ -87,33 +135,132 @@ def end_of_month(year, merch_month)
...
+      end
+    end
     
+    # Given beginning and end dates it will return an array of Fiscal Month's Start date
+    #
+    # @param start_date [Date] the starting date
+    # @param end_date [Date] the ending date
+    # @return [Array] Array of start dates of each Fiscal Month from given dates
     def merch_months_in(start_date, end_date)
+      merch_months_combos = merch_month_combo_from_dates(start_date, end_date)
```
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb]: Me and Masaki, are actually not sure about what is the point of this method?  Looking at older commits, and specs it seems like we need it to return the start dates of `merch_month`s that fall into and in between the `start_date` and `end_date`.

We asked Lisa S. who was the originator of this method, and she says that this method isn't used anymore.

File: spec/merch_calendar/retail_calendar_spec.rb (hunk 9 above)
Comment (chylauSF) [reply spec/merch_calendar/retail_calendar_spec.rb]: 👍

File: spec/merch_calendar/retail_calendar_spec.rb (hunk 10 above)
Comment (chylauSF) [reply spec/merch_calendar/retail_calendar_spec.rb]: 👍

File: spec/merch_calendar/retail_calendar_spec.rb (hunk 11 above)
Comment (chylauSF) [reply spec/merch_calendar/retail_calendar_spec.rb]: 👍

File: spec/merch_calendar/retail_calendar_spec.rb (hunk 12 above)
Comment (chylauSF) [reply spec/merch_calendar/retail_calendar_spec.rb]: Decided to delete the context block

**Comment by chylauSF** [comment]
Hi All,

First off thank you for your feedback!  Please continue to re-comb through this PR again, for more feedback.

I think I mainly concern with these things:
[ ] `merch_months_in` method where I don't know if we want this method any more, since it is causing more confusion in what it is actually doing, and I don't believe it is actually being used
[ ] The comments and documentation of the gem.
[ ] The ReadMe.md, to see if it is readable and understandable.

File: lib/merch_calendar/retail_calendar.rb (hunk 19 above)
[the remark this answers is not in the document]
Comment (jonworek) [reply lib/merch_calendar/retail_calendar.rb]: For the methods/comments that you modified, yes, you should.  If you didn't touch a method, you can leave it alone.

File: README.md [hunk 28]
```diff
@@ -21,15 +21,16 @@ gem "merch_calendar"
 
 ## Usage
 
-For converting a date into a `MerchWeek` object.
+For converting a date into a `MerchWeek` object.  `MerchWeek` object can take a date can translate to what merch week the date falls into.
```
Comment (jonworek) [inline README.md]: > object can take a date can translate 

I'm not sure what you were trying to say here, but it looks in need of revising.

File: README.md [hunk 29]
```diff
@@ -41,6 +42,9 @@ merch_week.end_of_month         # <Date: 2014-01-04>
 merch_week.start_of_year        # <Date: 2013-02-03>
 merch_week.end_of_year          # <Date: 2014-02-01>
 
+merch_week.calendar             # <MerchCalendar::RetailCalendar> (the calendar we're using)
```
Comment (jonworek) [inline README.md]: I don't know that I see a benefit to exposing this on the public interface of the class, but it doesn't really hurt anything.

File: README.md [hunk 30]
```diff
@@ -21,15 +21,16 @@ gem "merch_calendar"
...
 ```ruby
-merch_week = MerchCalendar::MerchWeek.from_date("2014-01-01")
+merch_week = MerchCalendar::MerchWeek.from_date("2014-01-01", calendar: MerchCalendar::RetailCalendar.new)
 
 merch_week.year                 # 2013 (the merch year associated with this date)
 merch_week.month                # 12 (the julian month that the date falls in)
 merch_week.week                 # 5 (the week number within the month)
-merch_week.year_week            # 48 (the week number within the year)
+merch_week.year_week            # 48 (the week number within the retail calendar year)
 merch_week.quarter              # 2
```
Comment (jonworek) [inline README.md]: Shouldn't this be `4`, since we're talking about a week in January?

File: lib/merch_calendar/merch_week.rb [hunk 31]
```diff
@@ -81,108 +93,96 @@ def year_week
...
     # @return [Date]
     def end_of_month
-      @end_of_month ||= retail_calendar.end_of_month(year, merch_month)
+      @end_of_month ||= calendar.end_of_month(year, merch_month)
     end
 
     # The merch season this date falls under.
     # Returns a string of +Fall/Winter+ or +Spring/Summer+
-    #
+    # THIS MIGHT NEED TO CHANGE DEPENDING ON THE CALENDAR
```
Comment (jonworek) [inline lib/merch_calendar/merch_week.rb]: Does this comment apply?  You seem to be changing it depending on the calendar.

File: lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb [hunk 32]
```diff
@@ -40,8 +50,29 @@ def start_of_quarter(year, quarter)
...
+      case merch_month
+      when 1,2,3
+        return QUARTER_1
+      when 4,5,6
+        return QUARTER_2
+      when 7,8,9
+        return QUARTER_3
+      when 10,11,12
+        return QUARTER_4
+      end
```
Comment (jonworek) [inline lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb]: It's a good practice to always have an `else` for your case statements.  Otherwise, the behavior is ambiguous. 

I generally raise an error if I trigger the `else` condition.

File: lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb [hunk 33]
```diff
@@ -87,33 +135,138 @@ def end_of_month(year, merch_month)
...
+    end
+    
+    # Given beginning and end dates it will return an array of Fiscal Month's Start date
+    #
+    # @param start_date [Date] the starting date
+    # @param end_date [Date] the ending date
+    # @return [Array] Array of start dates of each Fiscal Month from given dates
+    def merch_months_in(start_date, end_date) 
+      merch_months_combos = merch_year_and_month_from_dates(start_date, end_date)
```
Comment (jonworek) [inline lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb]: `merch_months_combos` is a collection of merch year and months?

File: spec/merch_calendar/retail_calendar_spec.rb [hunk 34]
```diff
@@ -121,5 +152,91 @@
...
+    end
+
+    it "raises an error for invalid merch months" do
+      expect { subject.julian_to_merch(13) }.to raise_error ArgumentError
+      expect { subject.julian_to_merch(0) }.to raise_error ArgumentError
+    end
+  end
+  
+  describe "#weeks_for_month" do
+    it "returns 4 weeks for a 4-week month Fiscal Year 2019 for Aug" do
```
Comment (jonworek) [inline spec/merch_calendar/retail_calendar_spec.rb]: I'm confused by the description.  Why are you referring to a "fiscal year" in the tests for the retail calendar?  Same in the test below.

Copy/paste error maybe?

File: spec/merch_calendar/stitch_fix_fiscal_year_calendar_spec.rb [hunk 35]
```diff
@@ -215,42 +244,121 @@
...
+      expect { subject.merch_to_julian(0) }.to raise_error ArgumentError
+    end
+  end
+
+  describe "#weeks_for_month" do
+    context "correct number of weeks given julian month and fiscal year" do
+      it "returns 4 weeks for a 4-week month Fiscal Year 2019 for Aug" do
+        weeks = subject.weeks_for_month(2019, 8)
+        expect(weeks.size).to eq 4
+      end
```
Comment (jonworek) [inline spec/merch_calendar/stitch_fix_fiscal_year_calendar_spec.rb]: Blank line needed between blocks.

File: lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb (hunk 32 above)
Comment (sleepn247) [reply lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb]: this would be consistent with the other raise logic too.

File: README.md (hunk 29 above)
Comment (chylauSF) [reply README.md]: What should I do? Should I delete the comment?

File: README.md (hunk 30 above)
Comment (chylauSF) [reply README.md]: yep! good catch! 👍

File: lib/merch_calendar/merch_week.rb (hunk 31 above)
Comment (chylauSF) [reply lib/merch_calendar/merch_week.rb]: Deleted! 👍

File: README.md (hunk 28 above)
Comment (chylauSF) [reply README.md]: I ended up deleting that sentence that I added.  I think the ruby text will speak for itself

File: lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb (hunk 32 above)
Comment (chylauSF) [reply lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb]: Should I type:
```ruby
    def quarter(merch_month)
      case merch_month
      when 1,2,3
        return QUARTER_1
      when 4,5,6
        return QUARTER_2
      when 7,8,9
        return QUARTER_3
      when 10,11,12
        return QUARTER_4
     else
        nil
      end
    end
    
```
OR
```ruby
    def quarter(merch_month)
      case merch_month
      when 1,2,3
        return QUARTER_1
      when 4,5,6
        return QUARTER_2
      when 7,8,9
        return QUARTER_3
     else
        return QUARTER_4
      end
    end
    
```

File: lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb (hunk 33 above)
Comment (chylauSF) [reply lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb]: yep, it is a collection of merch_year and merch_months, it helps with the `merch_months_in` in finding the `start_of_month` of each merch_month inside of the `merch_months_combos`

File: spec/merch_calendar/retail_calendar_spec.rb (hunk 34 above)
Comment (chylauSF) [reply spec/merch_calendar/retail_calendar_spec.rb]: yes it is! 👍

File: spec/merch_calendar/stitch_fix_fiscal_year_calendar_spec.rb (hunk 35 above)
Comment (chylauSF) [reply spec/merch_calendar/stitch_fix_fiscal_year_calendar_spec.rb]: 👍

File: lib/merch_calendar/retail_calendar.rb (hunk 19 above)
[the remark this answers is not in the document]
Comment (chylauSF) [reply lib/merch_calendar/retail_calendar.rb]: 👍

File: README.md (hunk 29 above)
Comment (jonworek) [reply README.md]: If it were me, I probably woudn't have added the `attr_accessor` to the class.  But I think it's fine to keep what you have.

File: lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb (hunk 32 above)
Comment (jonworek) [reply lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb]: neither :)

```
else
  raise "invalid merch month"
end
```

Comment (chylauSF) [reply lib/merch_calendar/stitch_fix_fiscal_year_calendar.rb]: got it!
